README.md
<h1 align="center">a11y-color-contrast</h1>
<p align="center">
    <a href="https://github.com/sondr3/a11y-color-contrast/actions"><img alt="GitHub Actions Status" src="https://github.com/sondr3/a11y-color-contrast/workflows/pipeline/badge.svg" /></a>
    <a href="https://www.npmjs.com/package/a11y-color-contrast"><img alt="npm" src="https://img.shields.io/npm/v/a11y-color-contrast" /></a>
    <a href="https://doc.deno.land/https://deno.land/x/a11y_color_contrast/mod.ts"><img src="https://img.shields.io/badge/deno-doc-blue?logo=deno"></a>
</p>

<p align="center">
    <b>A simple utility package for working with WCAG 2.2/3.0 color contrasts</b>
</p>

- **a11y**: Built for checking how readable colors are together
- **Simple:** Parsing of hex strings, contrast checks
- **WCAG**: Support for checking both WCAG 2.0 and WCAG 3.0 contrasts
- **APCA**: Support for the upcoming [APCA][apca] contrast algorithm

<details>
<summary>Table of Contents</summary>
<br />

- [Installation](#installation)
  - [Deno](#deno)
  - [Node](#node)
- [Getting started](#getting-started)
  - [`hex`](#hex)
  - [`wcag`](#wcag)
  - [`wcagContrastValue`](#wcagContrastValue)
  - [`wcagIsReadable`](#wcagIsReadable)
  - [`apcaContrastValue`](#apcaScore)
  - [`apcaToInterpolatedFont`](#apcaToInterpolatedFont)
  - [`apcaValidateFont`](#apcaValidateFont)
  - [Utilities](#utilities)
- [Inspiration and resources](#inspiration-and-resources)
- [License](#license)

</details>

# Installation

## Deno

```ts
import { apcaContrastValue, hex, wcagContrastValue } from "https://deno.land/x/a11y_color_contrast/mod.ts";

const wcag = wcagContrastValue(hex("#e1e1e1"), hex("#fff"));
const apca = apcaContrastValue(hex("#e1e1e1"), hex("#fff"));
```

## Node

Install the module with your favorite manager: `npm add a11y-color-contrast`

```ts ignore
import { apcaContrastValue, hex, wcagContrastValue } from "a11y-color-contrast";

const wcag = wcagContrastValue(hex("#e1e1e1"), hex("#fff"));
const apca = apcaContrastValue(hex("#e1e1e1"), hex("#fff"));
```

# Getting started

You can also read the documentation on
[Deno docs](https://doc.deno.land/https://deno.land/x/a11y_color_contrast/mod.ts).

## `hex`

`hex` is a utility function to parse a hex string to a `[number, number, number]`, useful if your colors are not in the
`RBG` format. Supports both short and long hex colors, and will strip out the alpha channel when the hex string contains
it. **Note,** the returned tripled will contain `NaN` if the string cannot be parsed. Use [`isValidColor`](#utilities)
to check if the input can be invalid.

```ts
// import { hex } from "a11y-color-contrast";
import { hex } from "https://deno.land/x/a11y_color_contrast/mod.ts";

hex("#fff");
hex("#e1e1e1");

// Also support hex4/hex8
hex("#1234");
hex("#11ff0000");

// And missing #
hex("fff");
```

## `wcag`

Based on the [WCAG 2.2][wcag] algorithm to calculate how readable two colors are when used together. The first argument
is the foreground color and the second the background. By default, the function defaults to checking whether the colors
pass the WCAG AAA standard (7:1 contrast ratio) for normal text.

```ts
// import { hex, wcag } from "a11y-color-contrast";
import { hex, wcag } from "https://deno.land/x/a11y_color_contrast/mod.ts";

wcag(hex("#fff"), hex("#e1e1e1"));
// { level: "AAA", size: "normal", score: 1.3076906134240802, pass: false }

wcag(hex("#0f0f0f"), hex("#fff"));
// { level: "AAA", size: "normal", score: 19.168645448127652, pass: true }

wcag(hex("#0f0f0f"), hex("#f4f"), { level: "AA" });
// { level: "AA", size: "normal", score: 6.8668010864317885, pass: true }

wcag(hex("#0f0f0f"), hex("#f4f"), { level: "AA", size: "large" });
// { level: "AA", size: "large", score: 6.8668010864317885, pass: true }
```

## `wcagContrastValue`

A simpler version of the [`wcag`](#wcag) function, this returns the contrast value between two colors based on the WCAG
2.2 algorithm.

```ts
// import { hex, wcagContrastValue } from "a11y-color-contrast";
import { hex, wcagContrastValue } from "https://deno.land/x/a11y_color_contrast/mod.ts";

wcagContrastValue(hex("#fff"), hex("#e1e1e1"));
// 1.3076906134240802

wcagContrastValue(hex("#0f0f0f"), hex("#fff"));
// 19.168645448127652
```

## `wcagIsReadable`

A simpler version of the [`wcag`](#wcag) function, this checks whether two colors used together are readable based on
the WCAG parameters passed.

```ts
// import { hex, wcagIsReadable } from "a11y-color-contrast";
import { hex, wcagIsReadable } from "https://deno.land/x/a11y_color_contrast/mod.ts";

wcagIsReadable(hex("#fff"), hex("#e1e1e1"));
// false

wcagIsReadable(hex("#0f0f0f"), hex("#fff"));
// true

wcagIsReadable(hex("#0f0f0f"), hex("#f4f"), { level: "AAA" });
// false

wcagIsReadable(hex("#0f0f0f"), hex("#f4f"), { level: "AA" });
// true
```

## `apcaContrastValue`

Based on the upcoming WCAG 3.0 standard, this function is based on the [APCA][apca-nut] algorithm to calculate how
readable two colors are when used together. The first argument is the foreground color and the second the background. It
is highly recommended reading the linked article and [resources](#inspiration-and-resources) to get an overview over the
differences between the WCAG and APCA standard.

```ts
// import { apcaContrastValue, hex } from "a11y-color-contrast";
import { apcaContrastValue, hex } from "https://deno.land/x/a11y_color_contrast/mod.ts";

apcaContrastValue(hex("#fff"), hex("#e1e1e1"));
// -17.5

apcaContrastValue(hex("#0f0f0f"), hex("#fff"));
// 105.5

apcaContrastValue(hex("#0f0f0f"), hex("#f4f"));
// 51.2
```

## `apcaToInterpolatedFont`

Based on a Lc between two colors, this will find the appropriate font sizes for it. The returned array will show
`"placeholder"` for when the contrast is too low for text and `"prohibited"` when the contrast is unusable and otherwise
a font size. If no font size can be calculated it will return null.

The returned array contains nine values, corresponding to the font useable at font weight 100 at index 0, and so on
until weight 900 at index 8.

```ts
// import { apcaToInterpolatedFont, hex } from "a11y-color-contrast";
import { apcaToInterpolatedFont, hex } from "https://deno.land/x/a11y_color_contrast/mod.ts";

apcaToInterpolatedFont(-17.5);
//  100 200 300 400 500 600 700 800 900
// [ "placeholder", ...]

apcaToInterpolatedFont(105.5);
//  100  200 300 400   500 600 700 800 900
// [ 39, 25, 18, 14.5, 14, 13, 12, 16, 18 ]

apcaToInterpolatedFont(51.2);
//  100  200 300 400 500 600   700   800   900
// [ 92, 69, 57, 31, 27, 23.5, 20.5, 20.5, 20.5 ]
```

## `apcaValidateFont`

Based on a Lc value, this function allows you to check whether a given font and weight combination passes the required
minimum contrast based on the APCA contrast table.

The first parameter is the Lc value, the second is either a single font size or an array of them and the third, optional
parameter is either a single font weight or an array of them. If the weight parameter is undefined, it will default to
all the font weights.

```ts
// import { apcaValidateFont, hex } from "a11y-color-contrast";
import { apcaValidateFont, hex } from "https://deno.land/x/a11y_color_contrast/mod.ts";

apcaValidateFont(-17.5, 36, 800);
// { "36": { "800": false } }

apcaValidateFont(105.5, [14, 16, 18], [400, 600, 800]);
// {
//   "14": { "400": true, "600": true, "800": false },
//   "16": { "400": true, "600": true, "800": true },
//   "18": { "400": true, "600": true, "800": true }
// }

apcaValidateFont(51.2, [18, 32]);
// {
//   "18": { "100": false, "200": false, ..., "700": false, "800": false, "900": false },
//   "32": { "100": false, "200": false, ..., "500": true, "600": true, "700": true, "800": true, "900": true }
// }
```

## Utility functions

### `toHex`

Converts an RGB triplet to its hex string representation.

```ts
// import { toHex } from "a11y-color-contrast";
import { toHex } from "https://deno.land/x/a11y_color_contrast/mod.ts";

toHex([0, 0, 0]);
// "#000000"
```

### `isValidColor`

Checks whether a color parsed via [`hex`](#hex) is valid.

```ts
// import { isValidColor } from "a11y-color-contrast";
import { isValidColor } from "https://deno.land/x/a11y_color_contrast/mod.ts";

isValidColor([0, 0, 0]);
// true

isValidColor([NaN, 0, 0]);
// false
```

### `colorFromObject`

Converts an RGB object into an RGB triplet.

```ts
// import { colorFromObject } from "a11y-color-contrast";
import { colorFromObject } from "https://deno.land/x/a11y_color_contrast/mod.ts";

colorFromObject({ r: 0, g: 0, b: 0 });
// [0, 0, 0]
```

# Inspiration and resources

- APCA homepage: https://git.apcacontrast.com/
- APCA calculator: https://www.myndex.com/APCA/
- APCA/WCAG calculator: https://www.myndex.com/BPCA/
- Contrast Calculator: https://contrast.tools/
- Accessible Palette: https://accessiblepalette.com/
- It’s time for a more sophisticated color contrast check for data visualizations:
  https://blog.datawrapper.de/color-contrast-check-data-vis-wcag-apca/

# License

MIT

[apca]: https://git.apcacontrast.com/
[apca-nut]: https://git.apcacontrast.com/documentation/APCA_in_a_Nutshell
[wcag]: https://www.w3.org/WAI/WCAG22/Techniques/general/G17.html


CHANGELOG.md
## v0.2.0

> 2022-07-19

## Summary

Fixed some unfortunate misspellings of WCAG, renamed a bunch of functions and added lots more documentation.

### Commits

- [[`2d95d74`](https://github.com/sondr3/a11y-color-contrast.git)] Yet another tiny mistake
- [[`105b82b`](https://github.com/sondr3/a11y-color-contrast.git)] Temporarily disable doc tests
- [[`a5f5ea9`](https://github.com/sondr3/a11y-color-contrast.git)] Update README
- [[`145b249`](https://github.com/sondr3/a11y-color-contrast.git)] Fix totally wrong values in apcaContrastValue docs
- [[`ad05e86`](https://github.com/sondr3/a11y-color-contrast.git)] Add documentation to isValidColor
- [[`12eb2e9`](https://github.com/sondr3/a11y-color-contrast.git)] Rename more things, more documentation
- [[`a0eebf0`](https://github.com/sondr3/a11y-color-contrast.git)] Create new wcagContrastValue function, more
  documentation
- [[`170b205`](https://github.com/sondr3/a11y-color-contrast.git)] Yeah, nah, no more grouped exports
- [[`99d110c`](https://github.com/sondr3/a11y-color-contrast.git)] And export the class
- [[`0027b63`](https://github.com/sondr3/a11y-color-contrast.git)] Fix names being wrong :facepalm:
- [[`bdb7b48`](https://github.com/sondr3/a11y-color-contrast.git)] Try with a class instead
- [[`75818aa`](https://github.com/sondr3/a11y-color-contrast.git)] Actually export wcag object
- [[`797f2a1`](https://github.com/sondr3/a11y-color-contrast.git)] Function renaming, add documentation
- [[`cfc5893`](https://github.com/sondr3/a11y-color-contrast.git)] Convert functions to named functions for deno doc
- [[`e9b3e07`](https://github.com/sondr3/a11y-color-contrast.git)] Add documentation, example to 'toHex' function
- [[`0d20fe6`](https://github.com/sondr3/a11y-color-contrast.git)] Change toHex method to use bit shifting dark magics
- [[`cfb3d73`](https://github.com/sondr3/a11y-color-contrast.git)] Run doctests on CI, fix README/examples so they
  compile
- [[`f53d244`](https://github.com/sondr3/a11y-color-contrast.git)] Fix wrong URL to documentation

## v0.1.0

> 2022-07-19

## Summary

Initial release of the package. See the README for usage notes.

### Commits

- [[`e8bfaf1`](https://github.com/sondr3/a11y-color-contrast.git)] Move node setup to top of pipeline
- [[`b46fd9a`](https://github.com/sondr3/a11y-color-contrast.git)] Publish job depends on the pipeline job
- [[`f27a99d`](https://github.com/sondr3/a11y-color-contrast.git)] Make publish step a separate CI job
- [[`b11d19d`](https://github.com/sondr3/a11y-color-contrast.git)] Run dnt on CI, add publishing step and tasks
- [[`5b25f03`](https://github.com/sondr3/a11y-color-contrast.git)] Fix links in README
- [[`1eaaefe`](https://github.com/sondr3/a11y-color-contrast.git)] Fix number 100 for optional, partial, default options
- [[`8023a6f`](https://github.com/sondr3/a11y-color-contrast.git)] Update README
- [[`84c3904`](https://github.com/sondr3/a11y-color-contrast.git)] Remove uneeded function, fix passing WCAG parameters
- [[`590de6c`](https://github.com/sondr3/a11y-color-contrast.git)] Remove redundant assert in tests
- [[`633ea4c`](https://github.com/sondr3/a11y-color-contrast.git)] Create utility function to validate a font against a
  Lc value
- [[`6cb7616`](https://github.com/sondr3/a11y-color-contrast.git)] Add documentation
- [[`97c6609`](https://github.com/sondr3/a11y-color-contrast.git)] Refactor and simplify some APCA functions
- [[`19c1267`](https://github.com/sondr3/a11y-color-contrast.git)] Add tests to the interpolated font lookup
- [[`9258c5a`](https://github.com/sondr3/a11y-color-contrast.git)] Implement fontLookupAPCA
- [[`35602d8`](https://github.com/sondr3/a11y-color-contrast.git)] Add contrast to font size table
- [[`12b3e0a`](https://github.com/sondr3/a11y-color-contrast.git)] Add function to get font size contrast
- [[`e2de2a6`](https://github.com/sondr3/a11y-color-contrast.git)] Refactor font contrast to a record object
- [[`9fd2cd2`](https://github.com/sondr3/a11y-color-contrast.git)] Add APCA font to contrast table
- [[`c061efe`](https://github.com/sondr3/a11y-color-contrast.git)] Rename generic 'Readability' type to 'WCAG'
- [[`7cc8203`](https://github.com/sondr3/a11y-color-contrast.git)] Fix some exporting names
- [[`ce84c99`](https://github.com/sondr3/a11y-color-contrast.git)] Fix formatting, convert deno.json -> deno.jsonc
- [[`fd7ca69`](https://github.com/sondr3/a11y-color-contrast.git)] Add dnt build tool for npm publishing
- [[`3105938`](https://github.com/sondr3/a11y-color-contrast.git)] Update README
- [[`8bb9e82`](https://github.com/sondr3/a11y-color-contrast.git)] Hide internal parsing functions
- [[`1f9a9d3`](https://github.com/sondr3/a11y-color-contrast.git)] Change Color type to be a tuple instead of object
- [[`0755453`](https://github.com/sondr3/a11y-color-contrast.git)] Remove WCAG 2.0/2.1 stuff
- [[`4a52171`](https://github.com/sondr3/a11y-color-contrast.git)] Add APCA contrast checking from WCAG3.0
- [[`6328844`](https://github.com/sondr3/a11y-color-contrast.git)] Add WCAG 2.0/2.1/2.2 changes to code
- [[`755ea00`](https://github.com/sondr3/a11y-color-contrast.git)] Extract WCAG stuff to their own file
- [[`7be066b`](https://github.com/sondr3/a11y-color-contrast.git)] Fix stupid misspelling of RGB
- [[`ad5d21d`](https://github.com/sondr3/a11y-color-contrast.git)] Start extracting functionality into small functions
- [[`318ffd0`](https://github.com/sondr3/a11y-color-contrast.git)] Extract parsing to its own file
- [[`30108be`](https://github.com/sondr3/a11y-color-contrast.git)] Port over color code from empyreum
- [[`f877b41`](https://github.com/sondr3/a11y-color-contrast.git)] Add basic RGB and RGBA functionality


build_npm.ts
import { build, emptyDir } from "https://deno.land/x/dnt@0.28.0/mod.ts";

await emptyDir("./npm");

await build({
  entryPoints: ["./mod.ts"],
  esModule: true,
  importMap: "./import_map.json",
  outDir: "./npm",
  shims: {
    deno: {
      test: "dev",
    },
  },
  compilerOptions: {
    target: "ES2021",
  },
  package: {
    name: "a11y-color-contrast",
    version: Deno.args[0]?.replace(/^v/, ""),
    author: "Sondre Nilsen <sondre@eons.io> (https://www.eons.io)",
    description: "A simple utility package for working with WCAG 2.2/3.0 color contrasts",
    license: "MIT",
    homepage: "https://github.com/sondr3/a11y-color-contrast#readme",
    repository: {
      type: "git",
      url: "git+https://github.com/sondr3/a11y-color-contrast.git",
    },
    bugs: {
      url: "https://github.com/sondr3/a11y-color-contrast/issues",
    },
    keywords: ["a11y", "wcag", "apca", "contrast", "color"],
    engines: {
      node: ">=16.6.0",
    },
  },
});

Deno.copyFileSync("LICENSE", "npm/LICENSE");
Deno.copyFileSync("README.md", "npm/README.md");
Deno.copyFileSync("CHANGELOG.md", "npm/CHANGELOG.md");
